### DISCUSSION TITLE
Tutorial on the types of tests to write for an R function

### DISCUSSION THREAD

**Discussion by jamesmbaazam** [epiverse-trace/epiverse-trace.github.io#282] [Resources]
### Context

Provides some guidance on how to write tests that cover four categories:

- Normal Cases: Typical input that should follow the main path through the code.
- Opposites: Cases that return opposite outputs (e.g. TRUE and FALSE).
- Large Cases: Typical input, but of a larger size than usual. This ensures that bugs don’t appear after multiple iterations.
- Edge Cases: Pairs of inputs that test different choice points in the code. For example, if a condition in the problem checks whether n < 2, two important edge cases are when n = 2 and n = 3, which trigger different behaviors. Other edge cases include the first / last characters in a string or items in a list.
- Special Cases: Some inputs need to be special-cased for many problems. This includes negative numbers, 0 and 1 for integers, the empty string (""), and input values of different types than are expected.

### URL

https://p4a.jhelvy.com/testing-debugging#test-case-types

### Suggested labels

Best Practice, Testing, Code coverage

**Comment by kylieainslie** [comment]
I think this is a great idea! I know this post is old, so maybe it's already gotten some attention. However, I'm happy to be a guinea pig for tutorial material. Would it also include using `testthat`?

**Comment by jamesmbaazam** [comment]
Hi @kylieainslie, I'm not sure I understand your comment "I'm happy to be a guinea pig for tutorial material. Would it also include using testthat?". Could you please expand on that? Are you looking to apply this resource to your own tests and need feedback?

**Reply by kylieainslie** [reply]
I just meant that if write a tutorial, I'm happy to go through it as a test case (no pun intended) - make sure the tutorial can be followed by someone who didn't write it. About testthat, I was asking whether you'd include examples of using testthat in the tutorial? Testing is not my strong suit, so I could benefit from a resource like this.

**Reply by jamesmbaazam** [reply]
Ah, thanks for explaining. I was only sharing the resource and did not intend to write a tutorial. If you were not aware of it, Hadley's R packages Book has a chapter on the [basics of unit testing with testthat](https://r-pkgs.org/testing-basics.html). I wonder if that would be helpful to you?

**Reply by kylieainslie** [reply]
Ah ok, I misunderstood!
